core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1979,6 +1979,38 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
   ): Stream[F2, O2] =
     that.mergeHaltL(this)
 
+  /** Given two sorted streams emits a single sorted stream, like in merge-sort.
+    * For entries that are considered equal by the Order, left stream element is emitted first.
+    * Note: both this and another streams MUST BE ORDERED already
+    * @example {{{
+    * scala> Stream(1, 2, 5, 6).interleaveOrdered(Stream(0, 2, 3, 4)).toList
+    * res0: List[Int] = List(0, 1, 2, 2, 3, 4, 5, 6)
+    * }}}
+    */
+  def interleaveOrdered[F2[x] >: F[x], O2 >: O: Order](that: Stream[F2, O2]): Stream[F2, O2] = {
+    val order = Order[O2].toOrdering // collections API needs Ordering, not cats.Order
+
+    def go(left: Pull[F2, O2, Unit], right: Pull[F2, O2, Unit]): Pull[F2, O2, Unit] =
+      left.uncons.flatMap {
+        case Some((lChunk, lNext)) =>
+          right.uncons.flatMap {
+            case Some((rChunk, rNext)) =>
+              val emitUpTo = order.min(lChunk(lChunk.size - 1), rChunk(rChunk.size - 1))
+              val emitLeftCount = lChunk.indexWhere(order.gt(_, emitUpTo)).getOrElse(lChunk.size)
+              val emitRightCount = rChunk.indexWhere(order.gt(_, emitUpTo)).getOrElse(rChunk.size)
+              val (emitLeft, keepLeft) = lChunk.splitAt(emitLeftCount)
+              val (emitRight, keepRight) = rChunk.splitAt(emitRightCount)
```

Comment:
Perhaps we should add a more elaborate function, one which operates in a single pass through each chunk and runs in linear time, as a single loop with two indices, one for each chunk chunks. 

That can be a matter for a subsequent PR, though. This one looks good enough.